src/balancer-strategy-manager/BalancerV2WeightedPoolStrategyManager.sol

```diff
@@ -0,0 +1,226 @@
+  function deposit(uint256[] calldata _tokenAmounts) external onlyOwnerOrGuardian {
+    if (_tokenAmounts.length != TOKEN_COUNT) {
+      revert TokenCountMismatch();
+    }
+
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      IERC20 token = IERC20(tokenConfig[i].token);
+
+      if (_tokenAmounts[i] > token.balanceOf(address(this))) {
+        revert InsufficientToken(address(token));
+      }
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    bytes memory userData = abi.encode(WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT);
+
+    IAsset[] memory assets = new IAsset[](TOKEN_COUNT);
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      IERC20(tokenConfig[i].token).safeIncreaseAllowance(address(VAULT), _tokenAmounts[i]);
+      assets[i] = IAsset(tokenConfig[i].token);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    userData = abi.encode(
+      WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT,
+      _tokenAmounts,
+      0
+    );
+
+    IBalancerVault.JoinPoolRequest memory request = IBalancerVault.JoinPoolRequest({
+      assets: assets,
+      maxAmountsIn: _tokenAmounts,
+      userData: userData,
+      fromInternalBalance: false
+    });
+
+    VAULT.joinPool(POOL_ID, address(this), address(this), request);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function withdraw(uint256 bpt) external onlyOwnerOrGuardian returns (uint256[] memory) {
+    return _withdraw(bpt);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function emergencyWithdraw() external onlyWithdrawable returns (uint256[] memory) {
+    uint256 bptAmount = IERC20(address(POOL)).balanceOf(address(this));
+
+    return _withdraw(bptAmount);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function getTokenConfig(uint256 id) external view override returns (TokenConfig memory) {
+    return tokenConfig[id];
+  }
+
+  /// @dev withdraws token from pool
+  function _withdraw(uint256 bptAmount) internal returns (uint256[] memory tokenAmounts) {
+    // IERC20(address(POOL)).safeIncreaseAllowance(address(VAULT), bptAmount);
+
+    uint256[] memory minAmountsOut = new uint256[](TOKEN_COUNT);
+    IAsset[] memory assets = new IAsset[](TOKEN_COUNT);
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      assets[i] = IAsset(tokenConfig[i].token);
+      minAmountsOut[i] = 0;
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    bytes memory userData = abi.encode(
+      WeightedPoolUserData.ExitKind.EXACT_BPT_IN_FOR_TOKENS_OUT,
+      bptAmount
+    );
+    IBalancerVault.ExitPoolRequest memory request = IBalancerVault.ExitPoolRequest({
+      assets: assets,
+      minAmountsOut: minAmountsOut,
+      userData: userData,
+      toInternalBalance: false
+    });
+
+    VAULT.exitPool(POOL_ID, address(this), payable(address(this)), request);
+
+    tokenAmounts = _backTokens();
+  }
+
+  /// @dev send back remaining tokens to provider
+  function _backTokens() internal returns (uint256[] memory) {
```

Comment:
The name of this function doesn't tell me anything, why is it called backToken if we're sending it to a provider?